### PR TITLE
sg: add "client codehost add-github" subcommand 

### PR DESCRIPTION
**Pull request by jhchabran**
The `sg client codehost add-github` command takes a given Sourcegraph instance and will remotely add code hosts as specified by the flags. I envisioned adding this to `mg` instead, but it was less straightforward as would need to deal with its yaml config, which is less "instant" to use compared to this. 

The original use case for this is to be able to easily add code hosts on the `scaletesting` instance. 

Possible things to do in the future: 

- Define a set of sourcegraph instances. In `sg.config.yaml` + secrets ? 
- Define "profiles" for codehosts, such as "large repo", "giga repo", "100k repos" for more convenient use? 

At this stage, let's keep things simple. 

See https://github.com/sourcegraph/handbook/pull/5131 for how it's being used in the docs. 

## Test plan

image

### CODE REVIEW

File: dev/sg/sg_client.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+						}
+						for _, s := range extSvcs {
+							if s.DisplayName == displayName {
+								std.Out.WriteSuccessf("Added CodeHost '%s'", displayName)
+								std.Out.WriteMarkdown(fmt.Sprintf("`%s`", s.Config))
+								return nil
+							}
+						}
+						std.Out.WriteFailuref("A code host named '%s' already exists with the following definition", displayName)
+						return errors.Wrap(err, "Could find newly added code host")
```
Comment (burmudar) [inline dev/sg/sg_client.go:87]:
```suggestion
						return errors.Wrapf(err, "Could not find newly added code host %q", displayName)
```

File: internal/gqltestutil/external_service.go [hunk 2]
```diff
@@ -121,3 +123,54 @@ mutation DeleteExternalService($externalService: ID!, $async: Boolean!) {
...
+				Nodes []ExternalServiceDefinition `json:"nodes"`
+			} `json:"externalServices"`
+		} `json:"data"`
+	}
+
+	q := query
+	var resp externalServicesResp
+	err := c.GraphQL("", q, nil, &resp)
+	if err != nil {
+		return nil, errors.Wrap(err, "request GraphQL")
```
Comment (burmudar) [inline internal/gqltestutil/external_service.go:173]:
```suggestion
		return nil, errors.Wrap(err, "GraphQL request")
```

File: dev/sg/client/codehost.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+package client
+
+import (
+	"encoding/json"
+
+	"github.com/sourcegraph/sourcegraph/internal/gqltestutil"
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+var GitHubHundredThousandsRepos = []string{
```
Comment (burmudar) [inline dev/sg/client/codehost.go:10]: This is not used?

File: dev/sg/sg_client.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+							return errors.Wrap(err, "Failed to sign-in on the Sourcegraph instance")
+						}
+						extSvcs, err := client.ExternalServices()
+						if err != nil {
+							return errors.Wrap(err, "Could not list code hosts")
+						}
+						displayName := cmd.String("display-name")
+						for _, s := range extSvcs {
+							if s.DisplayName == displayName {
+								std.Out.WriteFailuref("A code host named '%s' already exists with the following definition", displayName)
```
Comment (burmudar) [inline dev/sg/sg_client.go:65]:
```suggestion
								std.Out.WriteFailuref("A code host named '%s' already exists with the following name", displayName)
```

**Review by burmudar — APPROVED** [review]
Whooo! This is awesome and opens up a lot of possibilities!

File: dev/sg/client/codehost.go (hunk 3 above)
Comment (jhchabran) [reply dev/sg/client/codehost.go:10]: it's not, should be deleted, thanks

File: dev/sg/sg_client.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	Usage: "Perform common tasks on a Sourcegraph instance",
+	Subcommands: []*cli.Command{
+		{
+			Name:      "codehost",
+			ArgsUsage: " ",
+			Category:  CategoryUtil,
+			Usage:     "Interact with codehosts on a given Sourcegraph instance",
+			Subcommands: []*cli.Command{
+				{
+					Name:      "add-github",
```
Comment (sanderginn) [inline dev/sg/sg_client.go:24]: Perhaps we can split this into `add github`? Opens the door to `add gitlab` etc, as well as `remove` (if destructive operations are suitable for `sg`)

**Review by sanderginn — APPROVED** [review]

File: dev/sg/sg_client.go (hunk 5 above)
Comment (jhchabran) [reply dev/sg/sg_client.go:24]: Yeah, I was hesitating with this, so I went for the most straightforward solution until we got a better understanding of what we're building there.

**Comment by mrnugget** [comment]
I like the idea, but I'm also wondering: wouldn't this be better in `src-cli`? (To promote my idea again: we could have `sg src` run `src-cli` with env vars already setup, which would make everything added to `src-cli` also useful locally)

**Comment by jhchabran** [comment]
@mrnugget ah good call, I didn't think about adding this in `src-cli`. About having the ability to call other commands, yes, I thought about this many times, I think we're at the point where it's going to be really useful, especially with `mg` available as well.

**Comment by mrnugget** [comment]
Don't get me wrong: I like having utilities in `sg` that can help setup a local dev env faster, but since this is a GraphQL client _and_ it is helpful outside of local dev, I think it's better to have it in `src-cli` and then come up with another way on how to make that easily accessible locally (my `sg src` idea is just one way to do it, but (sorry, again) I think it would be super cool, because then you could also do `sg src users create` or `sg src api -query {'lol'}` easily)

**Comment by jhchabran** [comment]
@mrnugget we're 100% aligned, no worries 😊

**Comment by jhchabran** [comment]
Closed in favour of https://github.com/sourcegraph/src-cli/pull/856

**Pull request closed without merging**